#  WEEK 3 Segmenting and Clustering Neighborhoods in Toronto

In [1]:
# import related libraries
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
import requests
import json
from pandas.io.json import json_normalize 
from geopy.geocoders import Nominatim
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium
print('libraries stand by!')

libraries stand by!


##  Part 1 Generate the dataframe according to the Wiki page

In [2]:
# scrap data from wikipedia page
req = requests.get(" https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=945633050")
soup = BeautifulSoup(req.content,'lxml')
table = soup.find_all('table')[0]
df = pd.read_html(str(table))
neigh = pd.DataFrame(df[0])

#ignore cells with "Not assigned" borough coloums
neigh = neigh[~neigh["Neighbourhood"].isin(["Not assigned"])]
# rename column names for convenience
df_origin = neigh.rename(columns = {"Postcode": "PostalCode", "Neighbourhood": "Neighborhood"})
# combined rows which shared common postalcode and borough
df_origin = df_origin.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(lambda x:', '.join(x)).to_frame().reset_index()

#show the scraped dataframe
df_origin.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [3]:
# count the rows of dataframe
df_origin.shape[0]

103

Thus, we got 103 different neighborhoods

##  Part 2 Get the coordinates from the google map API

In [4]:
# the google map api was unconnectable, so I loaded the csv file directly.
coords = pd.read_csv("Geospatial_Coordinates.csv")
coords = coords.rename(columns = {"Postal Code": "PostalCode"})# rename for the sake of merging dfs
df_toronto = df_origin.merge(coords, on = "PostalCode", how = "left" )
df_toronto.sort_values('PostalCode')

PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, Martin Grove Gardens, Richv...  43.688905 -79.554724  
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437  
102                                          Northwest  43.706748 -79.594054  

[103 rows x 5 columns]

Thus, we got the ultimate toronto neighborhood data with coordinates

##  Part 3 Neighbourhood clustering in Toronto

Now, we try the clustering process in the lab.

In [5]:
# counts of the data
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_toronto['Borough'].unique()),
        df_toronto .shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [6]:
# define the user_agent, "TorontoDDD"
# generlizing the Toronto coordinates from the geocoder
address = 'Toronto'

geolocator = Nominatim(user_agent="TorontoDDD")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [7]:
# create a map of Toronto superimposed above using folium
Toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='orange',
        fill=True,
        fill_color='pink',
        fill_opacity=0.7,
        parse_html=False).add_to(Toronto_map)  

Toronto_map

In [8]:
"""
Scarborough_data = n_a2[n_a2['Borough'] == 'Scarborough'].reset_index(drop=True)
Scarborough_data
"""

"\nScarborough_data = n_a2[n_a2['Borough'] == 'Scarborough'].reset_index(drop=True)\nScarborough_data\n"

### Utilizing Foursqure API to explore the neighbourhoods and segment them

Here, I concealed the code of my credential for the Forsqure API in the next cell:

CLIENT_ID, CLIENT_SECRECR and VERSION='20180605'.

### Explore neighbourhoods in Toronto

Here, we skip showing the example of one single neighborhood and call the following functions directly to create the dataframe of vanues.

In [10]:
# functions from the Foursquare lab that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [11]:
#variables presetting
limit=100
radius=500

In [12]:
# function for collecting the vanues information through the Foursqure API using a for loop
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [13]:
# call the function to go through the toronto data set
Toronto_venues = getNearbyVenues(names = df_toronto['Neighborhood'],
                                   latitudes = df_toronto['Latitude'],
                                   longitudes = df_toronto['Longitude'])

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [14]:
# check the size of result
print(Toronto_venues.shape)
Toronto_venues.head()

(2146, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1                          Rouge, Malvern              43.806686   
2                          Rouge, Malvern              43.806686   
3  Highland Creek, Rouge Hill, Port Union              43.784535   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                         Venue  Venue Latitude  \
0              -79.194353                       Wendy’s       43.807448   
1              -79.194353         Interprovincial Group       43.805630   
2              -79.194353  T Hamilton & Son Roofing Inc       43.807985   
3              -79.160497         Royal Canadian Legion       43.782533   
4              -79.188711             G & G Electronics       43.765309   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.198194  Construction & Landscaping  
3       -79.163085                         Bar  
4       -79.191537           Electronics Store

In [15]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      9   
Alderwood, Long Branch                                                 10   
...                                                                   ...   
Willowdale West                                                         6   
Woburn                                                                  3   
Woodbine Gardens, Parkview Hill                                        11   
Woodbine Heights                                                       12   
York Mills West                                                         4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Agincourt                                                                5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       2   
Albion Gardens, Beaumond Heights, Humbergate, J...                       9   
Alderwood, Long Branch                                                  10   
...                                                                    ...   
Willowdale West                                                          6   
Woburn                                                                   3   
Woodbine Gardens, Parkview Hill                                         11   
Woodbine Heights                                                        12   
York Mills West                                                          4   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Agincourt                                               5               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...      2               2   
Albion Gardens, Beaumond Heights, Humbergate, J...      9               9   
Alderwood, Long Branch                                 10              10   
...                                                   ...             ...   
Willowdale West                                         6               6   
Woburn                                                  3               3   
Woodbine Gardens, Parkview Hill                        11              11   
Woodbine Heights                                       12              12   
York Mills West                                         4               4   

                                                    Venue Longitude  \
Neighborhood                                                          
Adelaide, King, Richmond                                        100   
Agincourt                                                         5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                2   
Albion Gardens, Beaumond Heights, Humbergate, J...                9   
Alderwood, Long Branch                                           10   
...                                                             ...   
Willowdale West                                                   6   
Woburn                                                            3   
Woodbine Gardens, Parkview Hill                                  11   
Woodbine Heights                                                 12   
York Mills West                                                   4   

                                                    Venue Category  
Neighborhood                            

In [16]:
print("There are {} uniques categories.".format(len(Toronto_venues["Venue Category"].unique())))

There are 262 uniques categories.


### Analyze Each Neighbourhood

In [17]:
# one hot coding and dataframe construction

T_onehot = pd.get_dummies(Toronto_venues[["Venue Category"]], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
T_onehot["Neighborhood"] = Toronto_venues["Neighborhood"] 

# move neighborhood column to the first column
fixed_columns = [T_onehot.columns[-1]] + list(T_onehot.columns[:-1])
T_onehot = T_onehot[fixed_columns]

T_onehot.head()


Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 262 columns]

In [18]:
T_onehot.shape

(2146, 262)

#### Group by neighbourhood

In [19]:
# utilizing the mean frequency of data to group neighbourhoods
T_grouped = T_onehot.groupby('Neighborhood').mean().reset_index()
T_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond          0.0   
1                                           Agincourt          0.0   
2   Agincourt North, L'Amoreaux East, Milliken, St...          0.0   
3   Albion Gardens, Beaumond Heights, Humbergate, ...          0.0   
4                              Alderwood, Long Branch          0.0   
..                                                ...          ...   
95                                    Willowdale West          0.0   
96                                             Woburn          0.0   
97                    Woodbine Gardens, Parkview Hill          0.0   
98                                   Woodbine Heights          0.0   
99                                    York Mills West          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1                 0.0                0.0      0.0                 0.0   
2                 0.0                0.0      0.0                 0.0   
3                 0.0                0.0      0.0                 0.0   
4                 0.0                0.0      0.0                 0.0   
..                ...                ...      ...                 ...   
95                0.0                0.0      0.0                 0.0   
96                0.0                0.0      0.0                 0.0   
97                0.0                0.0      0.0                 0.0   
98                0.0                0.0      0.0                 0.0   
99                0.0                0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
95           0.0             0.0              0.0               0.0  ...   
96           0.0             0.0              0.0               0.0  ...   
97           0.0             0.0              0.0               0.0  ...   
98           0.0             0.0              0.0               0.0  ...   
99           0.0             0.0              0.0               0.0  ...   

    Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0     0.0            0.0                           0.01               0.0   
1     0.0            0.0                           0.00               0.0   
2     0.0            0.0                           0.00               0.0   
3     0.0            0.0                           0.00               0.0   
4     0.0            0.0                           0.00               0.0   
..    ...            ...                            ...               ...   
95    0.0            0.0                           0.00               0.0   
96    0.0            0.0                           0.00               0.0   
97    0.0            0.0                           0.00               0.0   
98    0.0            0.0                           0.00               0.0   
99    0.0            0.0                           0.00               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000                    0.0              0.0       0.0   
1      0.000000                    0.0              0.0       0.0   
2      0.000000                    0.0              0.0       0.0   
3      0.000000                    0.0              0.0       0.0   
4      0.000000                    0.0              0.0       0.0   
..          ...                    ...              ...       ...   
95

In [20]:
T_grouped.shape

(100, 262)

#### top five common venues

In [21]:
num_top_venues = 5

for hood in T_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = T_grouped[T_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')




----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.09
1             Café  0.05
2       Restaurant  0.04
3              Gym  0.04
4  Thai Restaurant  0.03


----Agincourt----
                       venue  freq
0                     Lounge   0.2
1               Skating Rink   0.2
2             Breakfast Spot   0.2
3  Latin American Restaurant   0.2
4             Clothing Store   0.2


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                             venue  freq
0                       Playground   0.5
1                             Park   0.5
2                      Yoga Studio   0.0
3                    Metro Station   0.0
4  Molecular Gastronomy Restaurant   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0           Pizza Place  0.22
1         Grocery Store  0.22
2  Fast Food Restaurant  0.11
3              Pharmacy  0.11
4   

4                Café  0.08


----Highland Creek, Rouge Hill, Port Union----
                 venue  freq
0                  Bar   1.0
1          Yoga Studio   0.0
2  Moroccan Restaurant   0.0
3        Luggage Store   0.0
4               Market   0.0


----Hillcrest Village----
                      venue  freq
0        Athletics & Sports   0.2
1  Mediterranean Restaurant   0.2
2               Golf Course   0.2
3                      Pool   0.2
4                   Dog Run   0.2


----Humber Bay Shores, Mimico South, New Toronto----
         venue  freq
0  Coffee Shop  0.15
1          Gym  0.08
2       Bakery  0.08
3    Pet Store  0.08
4   Restaurant  0.08


----Humber Bay, King's Mill Park, Kingsway Park South East, Mimico NE, Old Mill South, The Queensway East, Royal York South East, Sunnylea----
                             venue  freq
0                             Park   0.5
1                   Baseball Field   0.5
2                      Yoga Studio   0.0
3                    Metro 

4  Mexican Restaurant  0.00


----The Kingsway, Montgomery Road, Old Mill North----
                             venue  freq
0                            River   0.5
1                             Pool   0.5
2                    Metro Station   0.0
3              Monument / Landmark   0.0
4  Molecular Gastronomy Restaurant   0.0


----Thorncliffe Park----
               venue  freq
0  Indian Restaurant  0.10
1        Yoga Studio  0.05
2       Intersection  0.05
3        Coffee Shop  0.05
4     Discount Store  0.05


----Victoria Village----
                   venue  freq
0           Hockey Arena  0.25
1            Coffee Shop  0.25
2  Portuguese Restaurant  0.25
3           Intersection  0.25
4         Medical Center  0.00


----Westmount----
                venue  freq
0         Pizza Place  0.29
1         Coffee Shop  0.14
2      Discount Store  0.14
3  Chinese Restaurant  0.14
4      Sandwich Place  0.14


----Weston----
                 venue  freq
0    Convenience Store   1.0
1    

#### put common venue data into df

In [22]:
# sort decending
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [23]:
# disply top 10 venues for each neighbourhoods

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = T_grouped['Neighborhood']

for ind in np.arange(T_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(T_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide, King, Richmond   
1                                          Agincourt   
2  Agincourt North, L'Amoreaux East, Milliken, St...   
3  Albion Gardens, Beaumond Heights, Humbergate, ...   
4                             Alderwood, Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop                  Café                   Gym   
1  Latin American Restaurant          Skating Rink                Lounge   
2                       Park            Playground         Women's Store   
3              Grocery Store           Pizza Place        Sandwich Place   
4                Pizza Place                   Gym           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Restaurant       Thai Restaurant                 Hotel   
1        Breakfast Spot        Clothing Store      Doner Restaurant   
2               Dog Run          Dessert Shop    Dim Sum Restaurant   
3            Beer Store  Fast Food Restaurant   Fried Chicken Joint   
4              Pharmacy          Skating Rink    Athletics & Sports   

  7th Most Common Venue    8th Most Common Venue 9th Most Common Venue  \
0         Deli / Bodega              Pizza Place           Salad Place   
1                 Diner           Discount Store   Distribution Center   
2                 Diner           Discount Store   Distribution Center   
3              Pharmacy  Comfort Food Restaurant          Dessert Shop   
4                   Pub                     Pool        Sandwich Place   

  10th Most Common Venue  
0                 Lounge  
1                Dog Run  
2             Donut Shop  
3    Dumpling Restaurant  
4                  Diner

### cluster neighbourhoods

In [24]:
# K means clustering to 5 clusters
# set number of clusters
kclusters = 5

T_grouped_clustering = T_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(T_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 4, 0, 0, 2, 2, 2, 2, 2])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
T_merged = df_toronto.merge(neighborhoods_venues_sorted, on='Neighborhood')
T_merged.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353               2            Print Shop  Fast Food Restaurant   
1 -79.160497               2                   Bar         Women's Store   
2 -79.188711               2     Electronics Store          Intersection   
3 -79.216917               2           Coffee Shop     Korean Restaurant   
4 -79.239476               2  Caribbean Restaurant           Gas Station   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Construction & Landscaping         Women's Store    Dim Sum Restaurant   
1                Dessert Shop  Ethiopian Restaurant   Empanada Restaurant   
2          Mexican Restaurant                  Bank        Medical Center   
3               Women's Store      Doner Restaurant    Dim Sum Restaurant   
4                      Bakery                  Bank    Athletics & Sports   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0                 Diner               Discount Store   Distribution Center   
1     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
2        Breakfast Spot          Rental Car Location             Drugstore   
3                 Diner               Discount Store   Distribution Center   
4       Thai Restaurant          Fried Chicken Joint      Hakka Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run             Donut Shop  
1             Drugstore             Donut Shop  
2            Donut Shop       Doner Restaurant  
3               Dog Run              Drugstore  
4            Donut Shop       Doner Restaurant

#### Visualize Clusters

In [28]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(T_merged['Latitude'], T_merged['Longitude'], T_merged['Neighborhood'], T_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Clusters

In [29]:
T_merged.loc[T_merged['Cluster Labels'] == 0, T_merged.columns[[1] + list(range(5, T_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22       North York               0         Grocery Store   
29       North York               0         Grocery Store   
31       North York               0         Grocery Store   
33        East York               0           Pizza Place   
42  Central Toronto               0                  Park   
70       North York               0                  Park   
79             York               0     Convenience Store   
87        Etobicoke               0           Pizza Place   
93       North York               0           Pizza Place   
96        Etobicoke               0           Pizza Place   
97        Etobicoke               0                  Park   
98        Etobicoke               0         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22                  Bank              Pharmacy           Pizza Place   
29                 Hotel                  Bank                  Park   
31    Athletics & Sports  Gym / Fitness Center          Liquor Store   
33        Breakfast Spot  Fast Food Restaurant          Intersection   
42           Swim School              Bus Line      Department Store   
70           Pizza Place                   Pub   Japanese Restaurant   
79         Grocery Store           Pizza Place              Bus Line   
87                   Gym           Coffee Shop              Pharmacy   
93   Empanada Restaurant         Women's Store               Dog Run   
96          Intersection           Coffee Shop        Discount Store   
97              Bus Line           Pizza Place        Sandwich Place   
98           Pizza Place        Sandwich Place            Beer Store   

     5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
22             Coffee Shop                 Home Service               Dog Run   
29  Gluten-free Restaurant             Department Store   Empanada Restaurant   
31              Donut Shop                        Diner        Discount Store   
33      Athletics & Sports                     Bus Line             Gastropub   
42       Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
70     Dumpling Restaurant                    Drugstore            Donut Shop   
79        Doner Restaurant           Dim Sum Restaurant                 Diner   
87            Skating Rink           Athletics & Sports                   Pub   
93            Dessert Shop           Dim Sum Restaurant                 Diner   
96      Chinese Restaurant               Sandwich Place         Women's Store   
97                 Dog Run           Dim Sum Restaurant                 Diner   
98    Fast Food Restaurant          Fried Chicken Joint              Pharmacy   

      8th Most Common Venue        9th Most Common Venue  \
22       Dim Sum Restaurant                        Diner   
29        Electronics Store  Eastern European Restaurant   
31      Distribution Center                      Dog Run   
33                     Bank                     Pharmacy   
42                Drugstore                   Donut Shop   
70         Doner Restaurant                      Dog Run   
79           Discount Store          Distribution Center   
87                     Pool               Sandwich Place   
93           Discount Store          Distribution Center   
96      Distribution Center           Dim Sum Restaurant   
97           Discount Store          Distribution Center   
98  Comfort Food Restaurant                 Dessert Shop   

   10th Most Common Venue  
22         Discount Store  
29    Dumpling Restaurant  
31       Doner Restaurant  
33   Gym / Fitness Center  
42       Doner Restaurant  
70           Dance Studio  
79                Dog Run  
87                  Diner  
93             Donut Shop  
96                  Diner  
97       Doner Restaurant  
98    Dumpling Restaurant

In [30]:
T_merged.loc[T_merged['Cluster Labels'] == 1, T_merged.columns[[1] + list(range(5, T_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61         Women's Store               Dog Run          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
61    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
61   Distribution Center      Doner Restaurant          Deli / Bodega

In [31]:
T_merged.loc[T_merged['Cluster Labels'] == 2, T_merged.columns[[1] + list(range(5, T_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0   Scarborough               2            Print Shop    Fast Food Restaurant   
1   Scarborough               2                   Bar           Women's Store   
2   Scarborough               2     Electronics Store            Intersection   
3   Scarborough               2           Coffee Shop       Korean Restaurant   
4   Scarborough               2  Caribbean Restaurant             Gas Station   
..          ...             ...                   ...                     ...   
91    Etobicoke               2           Golf Course           Women's Store   
92    Etobicoke               2                  Park             Pizza Place   
94   North York               2        Baseball Field  Furniture / Home Store   
95         York               2     Convenience Store           Women's Store   
99    Etobicoke               2   Rental Car Location               Drugstore   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Construction & Landscaping         Women's Store    Dim Sum Restaurant   
1                 Dessert Shop  Ethiopian Restaurant   Empanada Restaurant   
2           Mexican Restaurant                  Bank        Medical Center   
3                Women's Store      Doner Restaurant    Dim Sum Restaurant   
4                       Bakery                  Bank    Athletics & Sports   
..                         ...                   ...                   ...   
91            Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
92              Cosmetics Shop           Coffee Shop            Beer Store   
94               Women's Store    Dim Sum Restaurant                 Diner   
95            Department Store   Empanada Restaurant     Electronics Store   
99               Women's Store               Dog Run          Dessert Shop   

          6th Most Common Venue        7th Most Common Venue  \
0                         Diner               Discount Store   
1             Electronics Store  Eastern European Restaurant   
2                Breakfast Spot          Rental Car Location   
3                         Diner               Discount Store   
4               Thai Restaurant          Fried Chicken Joint   
..                          ...                          ...   
91                        Diner               Discount Store   
92                         Café               Shopping Plaza   
94               Discount Store          Distribution Center   
95  Eastern European Restaurant          Dumpling Restaurant   
99           Dim Sum Restaurant                        Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Distribution Center               Dog Run             Donut Shop  
1    Dumpling Restaurant             Drugstore             Donut Shop  
2              Drugstore            Donut Shop       Doner Restaurant  
3    Distribution Center               Dog Run              Drugstore  
4       Hakka Restaurant            Donut Shop       Doner Restaurant  
..                   ...                   ...                    ...  
91   Distribution Center               Dog Run             Donut Shop  
92          Liquor Store   Dumpling Restaurant              Drugstore  
94               Dog Run      Doner Restaurant       Department Store  
95             Drugstore            Donut Shop       Doner Restaurant  
99        Discount Store   Distribution Center       Doner Restaurant  

[78 rows x 12 columns]

In [32]:
T_merged.loc[T_merged['Cluster Labels'] == 3, T_merged.columns[[1] + list(range(5, T_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19  North York               3          Home Service         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19      Doner Restaurant    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19        Discount Store   Distribution Center               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
19            Donut Shop       Department Store

In [33]:
T_merged.loc[T_merged['Cluster Labels'] == 4, T_merged.columns[[1] + list(range(5, T_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough               4                  Park   
21        North York               4                  Park   
23        North York               4                  Park   
28        North York               4                  Park   
38         East York               4                  Park   
48  Downtown Toronto               4                  Park   
72              York               4                  Park   
89         Etobicoke               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
14            Playground         Women's Store                      Dog Run   
21     Convenience Store                  Bank                   Donut Shop   
23     Food & Drink Shop         Women's Store             Doner Restaurant   
28              Bus Stop               Airport             Doner Restaurant   
38     Convenience Store            Donut Shop           Dim Sum Restaurant   
48            Playground                 Trail  Eastern European Restaurant   
72         Women's Store                  Pool                      Dog Run   
89        Baseball Field         Women's Store             Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14          Dessert Shop    Dim Sum Restaurant                 Diner   
21                 Diner        Discount Store   Distribution Center   
23    Dim Sum Restaurant                 Diner        Discount Store   
28    Dim Sum Restaurant                 Diner        Discount Store   
38                 Diner        Discount Store   Distribution Center   
48   Dumpling Restaurant     Electronics Store             Drugstore   
72          Dessert Shop    Dim Sum Restaurant                 Diner   
89    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14        Discount Store   Distribution Center             Donut Shop  
21               Dog Run      Doner Restaurant          Women's Store  
23   Distribution Center               Dog Run             Donut Shop  
28   Distribution Center               Dog Run             Donut Shop  
38               Dog Run      Doner Restaurant          Women's Store  
48            Donut Shop      Doner Restaurant           Dance Studio  
72        Discount Store   Distribution Center             Donut Shop  
89   Distribution Center               Dog Run             Donut Shop